File: src/lib/scala/Slate.Common/src/main/scala/slate/common/databases/DbUtils.scala
/**
<slate_header>
  author: Kishore Reddy
  url: https://github.com/kishorereddy/scala-slate
  copyright: 2016 Kishore Reddy
  license: https://github.com/kishorereddy/scala-slate/blob/master/LICENSE.md
  desc: a scala micro-framework
  usage: Please refer to license on github for more info.
</slate_header>
  */

package slate.common.databases


import java.sql._
import scala.reflect.runtime.universe._

import slate.common.{DateTime, Files}


object DbUtils {

  /**
   * gets a new jdbc connection via Driver manager
 *
   * @return
   */
  def connect(con:DbCon) : Connection =
  {
    if(con.driver == "com.mysql.jdbc.Driver")
      DriverManager.getConnection(con.url, con.user, con.password)
    else
      DriverManager.getConnection(con.url)
  }


  /**
    * Execution template providing connection, statement with error-handling and connection closing
 *
    * @param con       : The connection string
    * @param callback  : The callback to call for when the connection is ready
    * @param error     : The callback to call for when an error occurrs
    */
  def execute(con:DbCon,
              callback:(Connection, Statement) => Unit,
              error:(Exception) => Unit): Unit =
  {
    val result = try
    {
      val c = connect(con)
      val s = c.createStatement()
      callback(c, s)
      (Option(c), Option(s))
    }
    catch
      {
        case ex:Exception =>
        {
          error(ex)
          (None, None)
        }
      }
    close(result._2, result._1)
  }


  /**
    * Execution template providing connection with error-handling and connection closing
 *
    * @param con       : The connection string
    * @param callback  : The callback to call for when the connection is ready
    * @param error     : The callback to call for when an error occurrs
    */
  def execute[T](con: DbCon, callback:(Connection) => T, error:(Exception) => Unit): Option[T] =
  {
    val result:(Option[Connection],Option[T]) = try
    {
      val c = connect(con)
      val r = callback(c)
      (Option(c), Option(r))
    }
    catch
      {
        case ex:Exception =>
        {
          error(ex)
          (None, None)
        }
      }
    close(result._1)
    result._2
  }


  /**
    * Execution template providing connection, statement with error-handling and connection closing
 *
    * @param con       : The connection string
    * @param callback  : The callback to call for when the connection is ready
    * @param error     : The callback to call for when an error occurrs
    */
  def executeStmt[T](con:DbCon,
              callback:(Connection, Statement) => T,
              error:(Exception) => Unit): Option[T] =
  {
    val result = try
    {
      val c = connect(con)
      val s = c.createStatement()
      val r = callback(c, s)
      (Option(c), Option(s), Option(r))
    }
    catch
    {
      case ex:Exception =>
      {
        error(ex)
        (None, None, None)
      }
    }
    close(result._2, result._1)
    result._3
  }


  /**
    * Execution template providing connection, prepared statement with error-handling & conn closing
 *
    * @param con       : The connection string
    * @param sql       : The sql text or stored proc name.
    * @param callback  : The callback to call for when the connection is ready
    * @param error     : The callback to call for when an error occurrs
    */
  def executePrepAs[T](con:DbCon,
              sql:String,
              callback:(Connection, PreparedStatement ) => T,
              error:(Exception) => Unit): Option[T] =
  {
    val result = try {
      val c = connect(con)
      val s = c.prepareCall(sql)
      val res = callback(c, s)
      (Option(c), Option(s), Option(res))
    }
    catch
      {
        case ex:Exception =>
        {
          error(ex)
          (None, None, None)
        }
      }
    close(result._2, result._1)
    result._3
  }


  /**
    * convenience function to fill prepared statement with parameters
 *
    * @param stmt
    * @param inputs
    */
  def fillArgs(stmt:PreparedStatement, inputs:Option[List[Any]]):Unit = {
    inputs.fold(Unit)( all => {
      all.indices.foreach( pos => {
        val arg = all(pos)
        arg match {
          case s:String         => stmt.setString (pos, arg.toString)
          case s:Int            => stmt.setInt    (pos, arg.asInstanceOf[Int])
          case s:Long           => stmt.setLong   (pos, arg.asInstanceOf[Long])
          case s:Double         => stmt.setDouble (pos, arg.asInstanceOf[Double])
          case s:Boolean        => stmt.setBoolean(pos, arg.asInstanceOf[Boolean])
          case s:DateTime       => "\"" + s.toString() + "\""
        }
      })
      Unit
    })
  }


  def getScalar[T](rs:ResultSet, typ:Type): Option[T] = {
    val pos = 1

    if(typ == typeOf[String] )       Option(rs.getString(pos).asInstanceOf[T])
    else if(typ == typeOf[Short]   ) Option(rs.getShort(pos).asInstanceOf[T])
    else if(typ == typeOf[Int]     ) Option(rs.getInt(pos).asInstanceOf[T])
    else if(typ == typeOf[Long]    ) Option(rs.getLong(pos).asInstanceOf[T])
    else if(typ == typeOf[Double]  ) Option(rs.getDouble(pos).asInstanceOf[T])
    else if(typ == typeOf[Boolean] ) Option(rs.getBoolean(pos).asInstanceOf[T])
    else if(typ == typeOf[DateTime]) Option(DateTime(rs.getTimestamp(pos)).asInstanceOf[T])
    else None
  }


  def close(rs: ResultSet):Unit = Option(rs).fold[Unit](Unit)( r => r.close() )


  def close(con: Connection):Unit = Option(con).fold[Unit](Unit)( c => c.close() )


  def close(con: Option[Connection]):Unit = con.fold[Unit](Unit)( c => c.close() )


  def close(stmt:Statement):Unit = Option(stmt).fold[Unit](Unit)( s => s.close() )


  def close(stmt: Statement, con:Connection): Unit = { close(stmt); close(con); }


  def close(stmt: Option[Statement], con:Option[Connection]): Unit = {
    stmt.foreach( s => close(s) )
    con.foreach( c => close(c) )
  }


  def close(rs:ResultSet, stmt: Statement, con:Connection): Unit =
  {
    close(rs)
    close(stmt)
    close(con)
  }


  /**
   * Loads a database config file containing connection properties from a users home directory.
   * The config file is a properties like file with key/value pairs
   *
   *  e.g. on windows {user_home}/.slate/db_default.txt
   *
   *   driver:com.mysql.jdbc.Driver
   *   url:jdbc:mysql://localhost/Database1
   *   user:root
   *   password:123456789
   *
   * @example         : DbConfig.loadFromUserFolder(".slate", "db_default.txt")
   * @param directory : directory name in the users home folder ( e.g. ".slate" )
   * @param fileName  : file name in the directory ( e.g. "db_default.txt" )
   * @return
   */
  def loadFromUserFolder(directory:String, fileName:String): DbConString =
  {
    val configResult = Files.readConfigFromUserFolder(directory, fileName)
    if(configResult.isEmpty)
      throw new IllegalArgumentException(s"database configuration file at $directory, $fileName not found")

    val config = configResult.get

    val con = new DbConString(
      config.getString("db.driver"),
      config.getString("db.url"),
      config.getString("db.user"),
      config.getString("db.password")
    )
    con
  }
}

File: src/lib/kotlin/slatekit-common/src/main/kotlin/slatekit/common/db/DbUtils.kt
/**
<slate_header>
author: Kishore Reddy
url: https://github.com/kishorereddy/scala-slate
copyright: 2016 Kishore Reddy
license: https://github.com/kishorereddy/scala-slate/blob/master/LICENSE.md
desc: a scala micro-framework
usage: Please refer to license on github for more info.
</slate_header>
 */

package slatekit.common.db


import slatekit.common.DateTime
import slatekit.common.Types
import java.sql.*
import java.time.*


object DbUtils {

    /**
     * gets a new jdbc connection via Driver manager
     *
     * @return
     */
    fun connect(con: DbCon): Connection =
            if (con.driver == "com.mysql.jdbc.Driver")
                DriverManager.getConnection(con.url, con.user, con.password)
            else
                DriverManager.getConnection(con.url)


    /**
     * Execution template providing connection with error-handling and connection closing
     *
     * @param con       : The connection string
     * @param callback  : The callback to call for when the connection is ready
     * @param error     : The callback to call for when an error occurrs
     */
    fun <T> executeCon(con: DbCon, callback: (Connection) -> T, error: (Exception) -> Unit): T? {
        val conn = connect(con)
        val result =
        try {
            conn.use { c ->
                callback(c)
            }
        } catch( ex:Exception ){
            error(ex)
            null
        }
        return result
    }


    /**
     * Execution template providing connection, statement with error-handling and connection closing
     *
     * @param con       : The connection string
     * @param callback  : The callback to call for when the connection is ready
     * @param error     : The callback to call for when an error occurrs
     */
    fun executeStmt(con: DbCon,
                    callback: (Connection, Statement) -> Unit,
                    error: (Exception) -> Unit): Unit {

        val conn = connect(con)
        try {
            conn.use { c ->
                val stmt = c.createStatement()
                stmt.use { s ->
                    callback(c, s)
                }
            }
        } catch( ex:Exception ){
            error(ex)
        }
    }


    /**
     * Execution template providing connection, prepared statement with error-handling & conn closing
     *
     * @param con       : The connection string
     * @param sql       : The sql text or stored proc name.
     * @param callback  : The callback to call for when the connection is ready
     * @param error     : The callback to call for when an error occurrs
     */
    fun <T> executePrepAs(con: DbCon,
                          sql: String,
                          callback: (Connection, PreparedStatement) -> T?,
                          error: (Exception) -> Unit): T? {

        val conn = connect(con)
        val result =
        try {
            conn.use { c ->
                val stmt = c.prepareCall(sql)
                stmt.use { s ->
                    val r = callback(c, s)
                    r
                }
            }
        } catch( ex:Exception ){
            error(ex)
            null
        }
        return result
    }

    /**
     * convenience function to fill prepared statement with parameters
     *
     * @param stmt
     * @param inputs
     */
    fun fillArgs(stmt: PreparedStatement, inputs: List<Any>?): Unit {
        inputs?.forEachIndexed { index, arg ->
            val pos = index + 1
            val jcls = arg.javaClass
            when (jcls) {
                Types.JStringAnyClass        -> stmt.setString(pos, arg.toString())
                Types.JBoolAnyClass          -> stmt.setBoolean(pos, arg as Boolean)
                Types.JShortAnyClass         -> stmt.setShort(pos, arg as Short)
                Types.JIntAnyClass           -> stmt.setInt(pos, arg as Int)
                Types.JLongAnyClass          -> stmt.setLong(pos, arg as Long)
                Types.JFloatAnyClass         -> stmt.setFloat(pos, arg as Float)
                Types.JDoubleAnyClass        -> stmt.setDouble(pos, arg as Double)
                Types.JLocalDateAnyClass     -> stmt.setDate(pos, java.sql.Date.valueOf(arg as LocalDate))
                Types.JLocalTimeAnyClass     -> stmt.setTime(pos, java.sql.Time.valueOf(arg as LocalTime))
                Types.JLocalDateTimeAnyClass -> stmt.setTimestamp(pos, java.sql.Timestamp.valueOf(arg as LocalDateTime))
                Types.JZonedDateTimeAnyClass -> stmt.setTimestamp(pos, java.sql.Timestamp.valueOf((arg as ZonedDateTime).toLocalDateTime()))
                Types.JInstantAnyClass       -> stmt.setTimestamp(pos, java.sql.Timestamp.valueOf(LocalDateTime.ofInstant(arg as Instant, ZoneId.systemDefault())))
                Types.JDateTimeAnyClass      -> stmt.setTimestamp(pos, java.sql.Timestamp.valueOf((arg as DateTime).local()))
            }
        }
    }


    @Suppress("UNCHECKED_CAST")
    fun <T> getScalar(rs: ResultSet, typ: Class<*>): T? {
        val pos = 1

        return if (typ == Types.JStringClass) rs.getString(pos) as T
        else if   (typ == Types.JBoolClass) rs.getBoolean(pos) as T
        else if   (typ == Types.JShortClass) rs.getShort(pos) as T
        else if   (typ == Types.JIntClass) rs.getInt(pos) as T
        else if   (typ == Types.JLongClass) rs.getLong(pos) as T
        else if   (typ == Types.JFloatClass) rs.getFloat(pos) as T
        else if   (typ == Types.JDoubleClass) rs.getDouble(pos) as T
        else if   (typ == Types.JLocalDateClass) rs.getDate(pos).toLocalDate() as T
        else if   (typ == Types.JLocalTimeClass) rs.getTime(pos).toLocalTime() as T
        else if   (typ == Types.JLocalDateTimeClass) rs.getTimestamp(pos).toLocalDateTime() as T
        else if   (typ == Types.JZonedDateTimeClass) rs.getTimestamp(pos).toLocalDateTime() as T
        else if   (typ == Types.JInstantClass) rs.getTimestamp(pos).toInstant() as T
        else if   (typ == Types.JDateTimeClass) DateTime.of(rs.getTimestamp(pos)) as T
        else null
    }


    fun getTypeFromLang(dataType: Class<*>):DbFieldType =
            if      (dataType == Types.JBoolClass    ) DbFieldTypeBool
            else if (dataType == Types.JStringClass  ) DbFieldTypeString
            else if (dataType == Types.JShortClass   ) DbFieldTypeShort
            else if (dataType == Types.JIntClass     ) DbFieldTypeNumber
            else if (dataType == Types.JLongClass    ) DbFieldTypeLong
            else if (dataType == Types.JDoubleClass  ) DbFieldTypeReal
            else if (dataType == Types.JLocalDateClass) DbFieldTypeLocalDate
            else if (dataType == Types.JLocalTimeClass) DbFieldTypeLocalTime
            else if (dataType == Types.JLocalDateTimeClass) DbFieldTypeLocalDateTime
            else if (dataType == Types.JZonedDateTimeClass) DbFieldTypeZonedDateTime
            else if (dataType == Types.JInstantClass) DbFieldTypeInstant
            else if (dataType == Types.JDateTimeClass) DbFieldTypeDateTime
            else DbFieldTypeString


    fun ensureField(text:String): String =
            if(text.isNullOrEmpty())
                ""
            else {
                text.trim().filter{ c -> c.isDigit() || c.isLetter() || c == '_'}
            }

}
